File: src/packages/control-ui/src/api/enhanced_geoip_lookup.ts
export interface GeoCidrRecord {
  netAddr: number;
  mask: number;
  countryCode: string;
}

export class EnhancedGeoIpLookup {
  private entries: GeoCidrRecord[] = [];

  addCidr(ipStr: string, maskBits: number, countryCode: string): void {
    const ipLong = this.ipToLong(ipStr);
    if (ipLong === null) return;
    const mask = maskBits === 0 ? 0 : (~0 << (32 - maskBits)) >>> 0;
    this.entries.push({
      netAddr: (ipLong & mask) >>> 0,
      mask,
      countryCode: countryCode.trim().toUpperCase(),
    });
  }

  lookup(ipStr: string): string | undefined {
    const ipLong = this.ipToLong(ipStr);
    if (ipLong === null) return undefined;

    let bestMatch: string | undefined = undefined;
    let bestMask = -1;

    for (const e of this.entries) {
      if ((ipLong & e.mask) >>> 0 === e.netAddr) {
        if (e.mask > bestMask) {
          bestMatch = e.countryCode;
          bestMask = e.mask;
        }
      }
    }
    return bestMatch;
  }

  static isPrivate(ipStr: string): boolean {
    const parts = ipStr.split('.').map((p) => parseInt(p, 10));
    if (parts.length !== 4) return false;
    const [first, second] = parts as [number, number];

    if (first === 10 || first === 127) return true;
    if (first === 172 && second >= 16 && second <= 31) return true;
    if (first === 192 && second === 168) return true;
    return false;
  }

  private ipToLong(ip: string): number | null {
    const parts = ip.split('.').map((p) => parseInt(p, 10));
    if (parts.length !== 4 || parts.some((p) => isNaN(p) || p < 0 || p > 255)) {
      return null;
    }
    return ((parts[0]! << 24) | (parts[1]! << 16) | (parts[2]! << 8) | parts[3]!) >>> 0;
  }
}

File: labs/mobile/donor-archive/tunnel/EnhancedGeoIpLookup.kt
package com.luminet.android.tunnel

data class AndroidGeoCidrEntry(
    val netAddr: Long,
    val mask: Long,
    val countryCode: String
)

class EnhancedGeoIpLookup {
    private val entries = mutableListOf<AndroidGeoCidrEntry>()

    fun addCidr(ipStr: String, maskBits: Int, countryCode: String) {
        val ipLong = parseIpv4ToLong(ipStr) ?: return
        val mask = if (maskBits == 0) 0L else (0xFFFFFFFFL shl (32 - maskBits)) and 0xFFFFFFFFL
        entries.add(
            AndroidGeoCidrEntry(
                netAddr = ipLong and mask,
                mask = mask,
                countryCode = countryCode.trim().uppercase()
            )
        )
    }

    fun lookup(ipStr: String): String? {
        val ipLong = parseIpv4ToLong(ipStr) ?: return null
        var bestMatch: String? = null
        var bestMask = -1L

        for (e in entries) {
            if ((ipLong and e.mask) == e.netAddr) {
                if (e.mask > bestMask) {
                    bestMatch = e.countryCode
                    bestMask = e.mask
                }
            }
        }
        return bestMatch
    }

    companion object {
        fun isPrivate(ipStr: String): Boolean {
            val parts = ipStr.split(".")
            if (parts.size != 4) return false
            val first = parts[0].toIntOrNull() ?: return false
            val second = parts[1].toIntOrNull() ?: return false

            return when (first) {
                10, 127 -> true
                172 -> second in 16..31
                192 -> second == 168
                else -> false
            }
        }

        private fun parseIpv4ToLong(ip: String): Long? {
            val parts = ip.split(".")
            if (parts.size != 4) return null
            var res = 0L
            for (p in parts) {
                val octet = p.toLongOrNull() ?: return null
                if (octet !in 0..255) return null
                res = (res shl 8) or octet
            }
            return res
        }
    }
}
